File: src/content/categories.py
"""
Category detection and mapping for kleinanzeigen.de.
"""

import json
import logging
from pathlib import Path
from typing import Optional, Tuple

logger = logging.getLogger(__name__)


class CategoryMapper:
    """Maps product information to kleinanzeigen.de categories."""
    
    def __init__(self, categories_file: Path):
        """
        Initialize the CategoryMapper.
        
        Args:
            categories_file: Path to categories.json file
        """
        with open(categories_file, 'r', encoding='utf-8') as f:
            self.data = json.load(f)
        
        self.categories = self.data['categories']
        self.keywords = self.data['keywords']
    
    def _match_keywords(self, text: str) -> Optional[str]:
        """Match text against category keywords."""
        text_lower = text.lower()
        
        best_match = None
        max_matches = 0
        
        for category, keywords in self.keywords.items():
            matches = sum(1 for keyword in keywords if keyword in text_lower)
            if matches > max_matches:
                max_matches = matches
                best_match = category
        
        return best_match if max_matches > 0 else None
    
    def _find_subcategory(self, category: str, text: str) -> Optional[str]:
        """Find best matching subcategory within a category."""
        if category not in self.categories:
            return None
        
        text_lower = text.lower()
        subcategories = self.categories[category]['subcategories']
        
        # Check each subcategory and its items
        for subcat, items in subcategories.items():
            # Check if subcategory name matches
            if subcat.lower() in text_lower:
                return subcat
            
            # Check if any items match
            for item in items:
                if item.lower() in text_lower:
                    return subcat
        
        # Return first subcategory as fallback
        return list(subcategories.keys())[0] if subcategories else None
    
    def map_category(self, product_name: str, product_description: str, 
                     detected_category: Optional[str] = None) -> Tuple[str, Optional[str]]:
        """
        Map product to kleinanzeigen.de category and subcategory.
        
        Args:
            product_name: Product name
            product_description: Product description
            detected_category: Category detected by vision analysis
            
        Returns:
            Tuple of (category, subcategory)
        """
        combined_text = f"{product_name} {product_description}"
        
        # Try to use detected category first
        category = None
        if detected_category:
            # Check if detected category matches our categories
            for cat in self.categories.keys():
                if cat.lower() in detected_category.lower() or detected_category.lower() in cat.lower():
                    category = cat
                    break
        
        # If no match, use keyword matching
        if not category:
            category = self._match_keywords(combined_text)
        
        # Fallback to "Sonstiges"
        if not category:
            category = "Sonstiges"
            logger.warning(f"Could not determine category, using fallback: {category}")
        
        # Find subcategory
        subcategory = self._find_subcategory(category, combined_text)
        
        logger.info(f"Mapped to category: {category}, subcategory: {subcategory}")
        return category, subcategory
    
    def get_all_categories(self) -> dict:
        """Get all available categories and subcategories."""
        return self.categories

File: src/content/categories.ts
/**
 * Category detection and mapping for kleinanzeigen.de.
 *
 * NOTE: This module is currently not used in the simplified workflow.
 * kleinanzeigen.de automatically detects categories from German titles.
 * This is kept for backwards compatibility and future use.
 */

import fs from 'fs';
import { createLogger } from '../utils/logger.js';

const logger = createLogger('CategoryMapper');

interface CategoryData {
  categories: Record<string, { subcategories: string[] }>;
  keywords: Record<string, string[]>;
}

/**
 * Maps product information to kleinanzeigen.de categories.
 */
export class CategoryMapper {
  private data: CategoryData;
  private categories: Record<string, { subcategories: string[] }>;
  private keywords: Record<string, string[]>;

  constructor(categoriesFile: string) {
    const fileContents = fs.readFileSync(categoriesFile, 'utf8');
    this.data = JSON.parse(fileContents) as CategoryData;
    this.categories = this.data.categories;
    this.keywords = this.data.keywords;
  }

  /**
   * Match text against category keywords.
   */
  private matchKeywords(text: string): string | null {
    const textLower = text.toLowerCase();

    let bestMatch: string | null = null;
    let maxMatches = 0;

    for (const [category, keywords] of Object.entries(this.keywords)) {
      const matches = keywords.filter((keyword) => textLower.includes(keyword)).length;
      if (matches > maxMatches) {
        maxMatches = matches;
        bestMatch = category;
      }
    }

    return maxMatches > 0 ? bestMatch : null;
  }

  /**
   * Find best matching subcategory within a category.
   */
  private findSubcategory(category: string, text: string): string | null {
    if (!(category in this.categories)) {
      return null;
    }

    const textLower = text.toLowerCase();
    const subcategories = this.categories[category].subcategories;

    for (const subcategory of subcategories) {
      if (textLower.includes(subcategory.toLowerCase())) {
        return subcategory;
      }
    }

    return null;
  }

  /**
   * Map product information to a category and subcategory.
   */
  mapCategory(name: string, description: string, detectedCategory?: string): [string, string | null] {
    logger.debug(`Mapping category for: ${name}`);

    // Combine name and description for matching
    const combinedText = `${name} ${description}`;

    // Try detected category first
    let category = detectedCategory;

    // If no detected category or not in our mapping, use keyword matching
    if (!category || !(category in this.categories)) {
      category = this.matchKeywords(combinedText) || undefined;
    }

    // Default to "Sonstiges" if no match
    if (!category) {
      category = 'Sonstiges';
      logger.info(`No category match found, defaulting to: ${category}`);
      return [category, null];
    }

    // Try to find subcategory
    const subcategory = this.findSubcategory(category, combinedText);

    logger.info(`Mapped to category: ${category}, subcategory: ${subcategory || 'None'}`);
    return [category, subcategory];
  }
}
